### PR TITLE
Google chrome add open tab feature

### PR DESCRIPTION
**Pull request by HenriChabrand**
## Description

Add "New Tab" command, allowing users to open a new tab with a search query. Currently opened or recently closed tabs matching the query will be suggested to avoid opening an existing tab.

Best use with `cmd + T` shortcut!

## Type of change

- Bug fix / improvement for somebody else's extension. Extension author: @ bromanko

## Screencast

Screenshot 2021-10-29 at 10 07 24

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I checked that extension's icon has size of `512x512`

### CODE REVIEW

File: extensions/google-chrome/src/google-chrome-new-tab.tsx [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+      activate
+      tell window 1
+          set newTab to make new tab ` + (queryText ? 'with properties {URL:"http://www.google.com/search?q='+queryText+'"}' : '') + ` 
+      end tell
+    end tell
+  `
+
+  const newTabOpened = await runAppleScript(script);
+
+  return newTabOpened;
```
Comment (pernielsentikaer) [inline extensions/google-chrome/src/google-chrome-new-tab.tsx:85]:
```suggestion
  return await runAppleScript(script);
```

The local variable is redundant here

File: extensions/google-chrome/src/google-chrome-new-tab.tsx [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+export default function Command(): ReactElement {
+    const [searchText, setSearchText] = useState<string>()
+    const { isLoading, error, entries } = useChromeHistorySearch(searchText)
+
+
+    const { useOriginalFavicon } = getPreferenceValues<{ useOriginalFavicon: boolean }>();
+    const [state, setState] = useState<State>({});
+
+    async function getTabs(query: string | null) {
+      var tabs = await getOpenTabs(useOriginalFavicon);
```
Comment (pernielsentikaer) [inline extensions/google-chrome/src/google-chrome-new-tab.tsx:115]:
```suggestion
      let tabs = await getOpenTabs(useOriginalFavicon);
```

File: extensions/google-chrome/src/google-chrome-new-tab.tsx [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    return `${this.windowsIndex}${Tab.TAB_CONTENTS_SEPARATOR}${this.tabIndex}`;
+  }
+
+  urlWithoutScheme(): string {
+    return this.url.replace(/(^\w+:|^)\/\//, '').replace('www.', '');
+  }
+
+  urlDomain(): string {
+    return this.urlWithoutScheme().split('/')[0];
+  }
```
Comment (pernielsentikaer) [inline extensions/google-chrome/src/google-chrome-new-tab.tsx:34]:
```suggestion
```

Remove this unused method

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @HenriChabrand 👋    Congrats on your first contribution! 👏 We are flattered that you chose the Raycast extensions repo for it.

*A few NITs and comments/suggestions before merging this beautifully gem* 💎 

- I don't think that "Open Tabs" should be in the list since it's already added to "Search Tabs" and it does only switch to the tab instead of opening a new tab 😄  

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by HenriChabrand** [comment]
Hello @pernielsentikaer 👋
Thanks for those feedbacks!

> * I don't think that "Open Tabs" should be in the list since it's already added to "Search Tabs" and it does only switch to the tab instead of opening a new tab 😄

I'm ust mimicking Chrome feature here:
Screenshot 2021-11-01 at 09 56 33

As a user, you don't always remember if the page you want to reach is already open in one of your tabs.
Having the `Open Tabs` section here prevents a user from opening multiple time the same page – even though he didn't remember that the page was already opened somewhere.

File: extensions/google-chrome/src/google-chrome-new-tab.tsx [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+
+async function openNewTab(queryText: string | null | undefined): Promise<boolean | string> {
+  popToRoot();
+  closeMainWindow({ clearRootSearch: true });
+  
+  const script = `
+    tell application "Google Chrome"
+      activate
+      tell window 1
+          set newTab to make new tab ` + (queryText ? 'with properties {URL:"http://www.google.com/search?q='+queryText+'"}' : '') + `
```
Comment (pernielsentikaer) [inline extensions/google-chrome/src/google-chrome-new-tab.tsx:75]:
```suggestion
          set newTab to make new tab ` + (queryText ? 'with properties {URL:"https://www.google.com/search?q='+queryText+'"}' : '') + ` 
```
HTTPS should be used here 🏎️

File: extensions/google-chrome/src/google-chrome-new-tab.tsx [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+    if (error) {
+        showToast(ToastStyle.Failure, "An Error Occurred", error.toString())
+    }
+
+    return (
+        <List onSearchTextChange={function (query) { setSearchText(query); getTabs(query);}} isLoading={isLoading} throttle={false}>
+            <List.Section title="New Tab" key="new-tab">
+              <List.Item
+                title={(!searchText ? "Open empty tab" : "Search \""+searchText+"\"")}
```
Comment (pernielsentikaer) [inline extensions/google-chrome/src/google-chrome-new-tab.tsx:132]:
```suggestion
                title={(!searchText ? "Open Empty Tab" : `Search "${searchText}"`)}
```

Corrected naming and added better readability

File: extensions/google-chrome/src/google-chrome-new-tab.tsx [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+
+const NewTabActions = (props: { query: string | undefined }): ReactElement => {
+    const query = props.query
+
+    return (
+        <ActionPanel title="New Tab">
+          <ActionPanel.Item
+            onAction={function () { openNewTab(query)}}
+            title={(query ? "Search \""+query+"\"" : "Open empty tab")} />
```
Comment (pernielsentikaer) [inline extensions/google-chrome/src/google-chrome-new-tab.tsx:160]:
```suggestion
            title={(query ? `Search "${query}"` : "Open Empty Tab")} />
```

Corrected naming and added better readability

File: extensions/google-chrome/src/google-chrome-new-tab.tsx [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+  );
+}
+
+function GoogleChromeGoToTab(props: { tab: Tab }) {
+  async function handleAction() {
+    await setActiveTab(props.tab);
+    await closeMainWindow();
+  }
+
+  return <ActionPanel.Item title="Open tab" icon={{ source: Icon.Eye }} onAction={handleAction} />;
```
Comment (pernielsentikaer) [inline extensions/google-chrome/src/google-chrome-new-tab.tsx:226]:
```suggestion
  return <ActionPanel.Item title="Open Tab" icon={{ source: Icon.Eye }} onAction={handleAction} />;
```

File: extensions/google-chrome/src/google-chrome-new-tab.tsx [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+}
+
+
+const HistoryItemActions = (props: { entry: HistoryEntry }): ReactElement => {
+    const { title, url } = props.entry
+
+    return (
+        <ActionPanel title={title}>
+          <OpenInBrowserAction
+              title="Open in tab"
```
Comment (pernielsentikaer) [inline extensions/google-chrome/src/google-chrome-new-tab.tsx:186]:
```suggestion
              title="Open in Tab"
```

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Just found a few more, when they are corrected I'll do the last review and make it ready to get merged - thanks for your update 💯 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by HenriChabrand** [comment]
Thank you so much for your review @pernielsentikaer 🙌

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @HenriChabrand 👋 

Just gave it a spin again, and it looks gorgeous. I did, however, find a thing that would be awesome to have addressed before merging.

It looks like the class Tab and a few functions are duplicated 1:1 from another file within the repo.

It makes sense to move it to utils with export instead, and import it respectively in the two locations in order to reduce repetition of software patterns, also known as [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) - which make bug fixing a lot easier.

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions or need help modifying it

Very so close to merging! 💯

**Comment by PitNikola** [comment]
@rgomezcasas, as an author of the original extension could you please take a look if you're okay with this addition?

**Comment by thomaspaulmann** [comment]
@HenriChabrand Is the Search list item always showing up?

**Comment by pernielsentikaer** [comment]
No @thomaspaulmann, default it starts like shown here

https://user-images.githubusercontent.com/38033107/140298407-ac8593e9-ddd5-44d6-a44d-c12a8e359bff.mp4

**Comment by HenriChabrand** [comment]
Hello @thomaspaulmann 👋

The first item is always here yes – but it will either be use to `Open an Empty Tab` or `Search` the term if some are inputted 👍

**Comment by thomaspaulmann** [comment]
> Hello @thomaspaulmann 👋
> 
> 
> 
> The first item is always here yes – but it will either be use to `Open an Empty Tab` or `Search` the term if some are inputted 👍 

Ah that makes sense. Also I didn't see that this is a separate command. Looking forward to it. I think it's a great addition.

**Comment by rgomezcasas** [comment]
Awesome!!

**Comment by pernielsentikaer** [comment]
Hi @HenriChabrand 👋 

Awesome contribution 💪 

A suggestion for later on could be if a domain is added (starts with www/http) then it would show "Go to" instead of search - this is not crucial to get it shipped but could be an awesome idea 💯

**Pull request merged**